src/compiler/moduleSpecifiers.ts

```diff
@@ -406,10 +401,25 @@ namespace ts.moduleSpecifiers {
         });
     }
 
-    function removeExtensionAndIndexPostFix(fileName: string, moduleResolutionKind: ModuleResolutionKind, addJsExtension: boolean): string {
+    function tryGetActualExtension(text: string, compilerOptions: CompilerOptions) {
+        const extension = pathIsRelative(text) && tryGetExtensionFromPath(text);
+        if (!extension) return undefined;
+
+        switch (extension) {
+            case Extension.Ts:
+                return Extension.Js;
+            case Extension.Tsx:
+                return compilerOptions.jsx === JsxEmit.React || compilerOptions.jsx === JsxEmit.ReactNative ? Extension.Js : Extension.Jsx;
+            default:
+                return extension;
+        }
+    }
+
+    function removeExtensionAndIndexPostFix(fileName: string, moduleResolutionKind: ModuleResolutionKind, compilerOptions: CompilerOptions, preferences: ModuleSpecifierPreferences): string {
         const noExtension = removeFileExtension(fileName);
-        return addJsExtension
-            ? noExtension + ".js"
+        const actualExtension = tryGetActualExtension(fileName, compilerOptions);
+        return (preferences.includeExtensionInImports && actualExtension)

```

Comment:
If `includeExtensionInImports` is `undefined` (so not explicitly set), we should still be doing what the `usesJsExtensionOnImports` function did - context based method of determining which way we should go based on what's already there.